### ISSUE TITLE
[obsolete] WIP

### ISSUE THREAD

**Issue by gitvasile** [ottokiksmaler/nx500_nx1_modding#23]
**active** means in active development. Otherwise there was some time spent on it, but no active development/effort. List is **not** ordered. No ETA and no result guarantee.
- [ ] **burner cam received** - **active** - better way to launch mods from hibernation, with override on boot with key pressed
- [ ] remove hardcoded time limits for NX500
- [ ] **active** integrate keyscan into hibernate image - [here](https://github.com/ottokiksmaler/nx500_nx1_modding/blob/master/doc/Update_hibernation_image.md)
- [ ] rsync backup when connected to home wifi
- [ ] uncropped 4K for NX500
- [ ] video LUTs
- [ ] break into T-Kernel to see if more video modes are feasible
- [ ] disable video NR - update: good chance that video NR is already factory disabled

**Comment by ottokiksmaler** [comment]
I can do rsync when I find time. I have to first test and finish the time lapse app I just made and then I'm on the ssh and rsync.

**Comment by gitvasile** [comment]
I _just_ finished compiling rsync. :-) - needed it because I am tired of pulling the sd card to copy files on the camera. Will upload it shortly :-)

**Comment by gitvasile** [comment]
rsync available in master branch.

**Comment by ottokiksmaler** [comment]
Excellent for rsync, I will need ssh as well for it, so I'll try my luck with openssh. 
Regarding compiling - you need files from Samsung open source repository (ones with lib and usr) but here most likely an unpacked rootfs where it can be found (maybe over the os files or add it manually with rpath) as you are using libs from it.

**Comment by ottokiksmaler** [comment]
Also, all missing sem_ functions are for semaphore handling (they should be in glibc on the camera rootfs l

**Comment by gitvasile** [comment]
you don't need ssh for rsync. Just set up rsync as a service on your PC and connect to it from the camera. Give me a shout if you need info.

**Comment by gitvasile** [comment]
can't find glibc on nx500...

**Comment by ottokiksmaler** [comment]
I know I don't need it but I wanted to use it when I'm on a public wifi or similar to upload images to server. 
Glibc should be at /usr/lib/libc.so or .a for static.

**Comment by gitvasile** [comment]
@ottokiksmaler ok I give up. Can't make it compile and run properly. When you get around to calling functions with a program on your own, please let me know so I can get a copy and replace with what I need. :-( thx

**Comment by ottokiksmaler** [comment]
Will do, but there's no guarantee it will work at all.

**Comment by ottokiksmaler** [comment]
Ho, Ho, Ho, here comes Santa and he's bringing in the presents ;)

TLDR: It works, you just need to put libprefman.so in your compile dir :)

Longer version:

Notes:
~/mnt is where the rootfs is mounted ../arm/ is where the packages from Samsung open source are unpacked (../arm/lib ../arm/usr)

Go to your directory where the source is, then
```
cp ~/mnt/usr/lib/libprefman.so ./
arm-linux-gnueabihf-gcc -o callsta callsta.c --sysroot=../arm/ libprefman.so -Wl,-dynamic-linker,/lib/ld-2.13.so
```
That's it.

But, here is mine callsta.c (I went with pref_get_long as I know what to expect and where)
```
#include <stdio.h>

long int pref_get_long(int pref_id, int pref_offset);

int main(int argc, const char *argv[])
{
    printf("Hello from main!\n");
    printf("pref_get_long(3,0x208)=%ld\n", pref_get_long(3,0x208));
    return 0;
}
```

Copied it over to camera and:
```
[root@drime5 ~]# prefman get 3 0x208 l                              
[sysrw] in memory: 


 value = 5846 (0x000016d6) 

[root@drime5 ~]# /mnt/mmc/callsta          
Hello from main!
pref_get_long(3,0x208)=5846
```

It works :)

Now have fun with it, I was planning on trying to compile against capture lib to send it directly what I want but I have almost no time next few days.

**Comment by ottokiksmaler** [comment]
Documented it here https://github.com/ottokiksmaler/nx500_nx1_modding/blob/master/doc/Compile_against_camera_library.md

**Comment by gitvasile** [comment]
You do realize that using this we will be able to take over the camera and operate it in parallel with di-camera-app, right? i.e. we would be able to basically have a parallel UI if we want to :-).

Assuming, of course, that we get to understand the various functions offered by the shared libraries.

**Comment by ottokiksmaler** [comment]
We do have to take care of locking and semaphores or our programs will fail (it's shared between rtos and us). Also, we have to communicate with di-camera-app or push it aside not to interfere. 
But, these are all solvable issues. 
Low hanging fruit is to batch operations that now take several calls to st and prefman - this will both increase performance and reliability.

**Comment by ottokiksmaler** [comment]
I actually wanted to do this for a long time but I assumed it would me much harder to pull off.

**Comment by gitvasile** [comment]
I will start to look through my notes to see some interesting APIs I noted in passing...

**Comment by ottokiksmaler** [comment]
Mmf-camcorder and libcapture-fw-prod or whatever the name is seem most interesting

**Comment by Tommixoft** [comment]
Why don't you create website for NX hacks? i can help if you want [can do anything from a to z what relates to website stuff]. Maybe having a website everything will look more stable and professional...Maybe it will draw more attention. 
Only need neat name for domain :) NXhacks.com teken :D
NxLantern or LanterNX (sounds more like product name) ;D Helping DPreview to get more visits because of your work - is not an option.

**Comment by ottokiksmaler** [comment]
I don't know, I kinda like how github works, allowing easy access for everybody, including forking, etc.

**Comment by Tommixoft** [comment]
But GitHab is not representative... for devs maybe..but not for those you should be interested in. 
Maybe it's early now, but still.. You two doing work, and in my opinion deserve to be a little bit more known than just on dpreview. Yes you do not cure cancer, but still..magicLantern too was small some time ago.. sad part is that nx we know maybe will never be alive again..so this all project is doomed to die anyways :( Unless Samung will see potential and bring nx line again, like in the past.

**Comment by gitvasile** [comment]
Github also has a wiki, easy to do something nice with it.

**Comment by gitvasile** [comment]
di-camera-app is basically idle unless you press a button on the camera, and we can easily intercept these to stop them from getting to the di-camera-app.

**Comment by ottokiksmaler** [comment]
Intercepting is trivial - just open a window and it grabs all the events.

**Comment by gitvasile** [comment]
Cannot wait to get the burner NX500 :-). Hibernation image updating  + /opt mod storage + fast emulation of st calls by direct library calls should permanently solve the performance problems :-)

**Comment by ottokiksmaler** [comment]
Just tried to compile code that uses libmm-camcorder but here adapted for libmmf ones and it's a mess - a real whack-a-mole of function names and library symbols. Will keep trying tomorrow but does not look promising right now.

**Comment by gitvasile** [comment]
gofundme NX500 FUNDED!

I just ordered the camera. :-)

hibernation image, here i come.

**Comment by ottokiksmaler** [comment]
CONGRATULATIONS! 

Now comes the hard part - waiting for it to arrive ;)

**Comment by gitvasile** [comment]
well, thanks go for the most part to ttbek@dpreview - he contributed like >400 EUR!!!

**Comment by ottokiksmaler** [comment]
Wow, that's quite a donation :)

**Comment by Tommixoft** [comment]
Wow in my country minimal monthly salary is 325 eur. :D

2016-05-07 17:01 GMT+03:00 gitvasile <notifications@github.com>:

**Comment by ge0rg** [comment]
In case anybody is still watching, there is now SSH support! :dancers: 

I've compiled Dropbear (very tiny) and OpenSSH (for SFTP and SCP), and put the binaries and installer (as well as docs on how to compile) under https://github.com/ottokiksmaler/nx500_nx1_modding/tree/master/ssh-server
